contributors/design-proposals/preserve-order-in-strategic-merge-patch.md

```diff
@@ -0,0 +1,547 @@
+
+One use case is the environment variables. We don't preserve the order which causes
+issue [40373](https://github.com/kubernetes/kubernetes/issues/40373).
+
+## Proposed Change
+
+Changes are all in strategic merge patch package.
+The proposed solution is similar to the solution used for deleting elements from lists of primitives.
+
+Add to the current patch, a directive ($setElementOrder) containing a list of element keys -
+either the patch merge key, or for primitives the value. When applying the patch,
+the server ensures that the relative ordering of elements matches the directive.
+
+Here is an simple example:
+
+Suppose we have a type called list. The patch will look like below.
+The order from the parallel list ($setElementOrder/list) will be respected.
+
+```yaml
+$setElementOrder/list:
+- A
+- B
+list:
+- B
+- A
+```
+
+All the items in the server's live list but not in the parallel list will be append to the end of the parallel list.
+The relative order between these appended items are kept.
+If the relative order of the live config in the server is different from the order of the parallel list,
+the user's patch will always override the order in the server.
+
+The patched list will look like:
+
+```
+mergingList:
+- parallelListItem1 \
+  ...                |===> items from the parallel list
+- parallelListItemN /
+- otherItem1 \
+  ...         |===> items in the server's list but not in the parallel list
+- otherItemN /
+```
+
+### When $setElementOrder is not present and patching a list
+
+The new directive $setElementOrder is optional.
+The new server applying a patch request without the directive will be
+a best-effort operation to keep the order of elements that are already in the list:
+
+For changes and deletions, the server will just merge the change without sorting them.
+So the items in the list will retain the order.
+
+For additions, the items will be appended to the end of the list.
+
+So it changes a little from previous releases, but is still fully backward compatible with old clients.
+
+Examples where A and C have been changed, B has been deleted and D has been added.
+
+Patch:
+
+```yaml
+list:
+- A'
+- $patch: delete
+  B
+- D
+```
+
+Live:
+
+```yaml
+list:
+- C
+- B
+- A
+```
+
+Result:
+
+```yaml
+list:
+- C
+- A
+- D
+```
+
+### `$setElementOrder` may contain elements not present in the patch list
+
+The $setElementOrder value may contain elements that are not present in the patch
+but present in the list to be merge to reorder the elements as part of the merge.
+
+Example where A & B have not changed:
+
+Patch:
+
+```yaml
+$setElementOrder/list:
+- A
+- B
+```
+
+Live:
+
+```yaml
+list:
+- B
+- A
+```
+
+Result:
+
+```yaml
+list:
+- A
+- B
+```
+
+### When the list to be merged contains elements not found in `$setElementOrder`
+
+If the list to be merged contains elements not found in $setElementOrder,
+they will come after all elements defined in $setElementOrder, but keep their relative ordering.
```

Comment:
I think @liggitt was suggesting both are required.  I think that is probably fine.